### PR TITLE
Oc 8481:Add PDF Casebook functionality for one form

### PR DESCRIPTION
**Pull request by Taoli2018**

### CODE REVIEW

File: web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java [hunk 1]
```diff
@@ -503,4 +523,84 @@ public String startBulkAddParticipantJob(MultipartFile file, String schema, Stri
 		});
 		return jobDetail.getUuid();
 	}
+	
+	@ApiOperation(value = "To get PDF version casebook for one specific participant",  notes = "only work for authorized users with the right acecss permission ")
+	@RequestMapping(value = "/pdf/print/{studyOid}/{studySubjectIdentifier}", method = RequestMethod.POST)
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java]: If participantId is unique within a study, let's use this url: /studies/{studyOID}/participants/{participantId}/casebook

File: web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java [hunk 2]
```diff
@@ -503,4 +523,84 @@ public String startBulkAddParticipantJob(MultipartFile file, String schema, Stri
...
+			  logger.info("REST request to Casebook PDF Job uuid {} ", uuid);			
+			  return new ResponseEntity<Object>("job uuid: " + uuid, HttpStatus.OK);
+		  
+			  } catch (OpenClinicaSystemException e) {
+					return new ResponseEntity(validateService.getResponseForException(e, studyOid, siteOid), HttpStatus.BAD_REQUEST);
+			}
+		 	 
+		 }
+
+	@Transactional
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java]: we don't need @Transactional here

File: web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java [hunk 3]
```diff
@@ -375,6 +409,198 @@ private boolean isEnrollmentCapEnforced(StudyBean currentStudy,StudyBean siteStu
...
+        return finalFile;
+     }
+
+
+	/**
+	 * @param studyOID
+	 * @param studySubjectIdentifier
+	 * @return
+	 */
+	public String getMergedPDFcasebookFileName(String studyOID, String participantId) {
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java]: move this to controller?

File: web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java [hunk 4]
```diff
@@ -503,4 +523,84 @@ public String startBulkAddParticipantJob(MultipartFile file, String schema, Stri
...
+			jobDetail.setLogPath(fileName);
+			ServletContext servletContext = request.getServletContext();
+			String accessToken = (String) request.getSession().getAttribute("accessToken");
+			servletContext.setAttribute("accessToken", accessToken);
+			servletContext.setAttribute("studyID", study.getStudyId()+"");
+			List<String> permissionTagsString =permissionService.getPermissionTagsList((StudyBean)request.getSession().getAttribute("study"),request);
+			CompletableFuture<Object> future = CompletableFuture.supplyAsync(() -> {
+				try {
+					 String userAccountID = userAccountBean.getId() +"";
+					 File pdfFile = this.studyParticipantService.startCaseBookPDFJob(jobDetail,studyOid, studySubjectIdentifier, servletContext, userAccountID, fullFinalFilePathName,format, margin, landscape,permissionTagsString);
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/controller/StudyParticipantController.java]: this doesn't need to return the file

File: web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java [hunk 5]
```diff
@@ -375,6 +409,198 @@ private boolean isEnrollmentCapEnforced(StudyBean currentStudy,StudyBean siteStu
...
+						            ServletContext servletContext,
+						            String userAccountID,                    
+						            String fullFinalFilePathName,
+						            String format, 
+						            String margin, 
+						            String landscape,
+						            List<String> permissionTags) throws Exception {
+		
+    	    ArrayList<File> pdfFiles = new ArrayList<File>();
+		    File mergedPdfFile = null;
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java]: fix indentation

File: web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java [hunk 6]
```diff
@@ -375,6 +409,198 @@ private boolean isEnrollmentCapEnforced(StudyBean currentStudy,StudyBean siteStu
...
+			    	List<EventCrf> eventCRFs = studyEvent.getEventCrfs();
+			    	
+			    	for(EventCrf eventCrf : eventCRFs) {
+			    		formLayoutOID = eventCrf.getFormLayout().getOcOid();
+			    		
+			    		int studyEventDefinitionId = studyEvent.getStudyEventDefinition().getStudyEventDefinitionId();
+			    		EventDefinitionCrf edc = eventDefinitionCrfDao.findByStudyEventDefinitionIdAndCRFIdAndStudyId(studyEventDefinitionId, eventCrf.getCrfVersion().getCrf().getCrfId(), studyId);
+			    	    if(edc != null && validateService.hasCRFpermissionTag(edc, permissionTags)) {
+			    	    	PFormCacheSubjectContextEntry subjectContext = new PFormCacheSubjectContextEntry();
+				    		studyEventDefinitionID = studyEventDefinitionId + "";
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java]: fix indentation

File: web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java [hunk 7]
```diff
@@ -375,6 +409,198 @@ private boolean isEnrollmentCapEnforced(StudyBean currentStudy,StudyBean siteStu
...
+    }
+    /**
+     * AC5: The PDF Casebook will be called "Participant <Participant ID> Casebook <current timestamp>.pdf".
+     * @param files
+     * @param studyOID
+     * @param studySubjectIdentifier
+     * @return
+     * @throws IOException
+     */
+    public File mergePDF(ArrayList<File> files,
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java]: move this method to a new service called "PdfService" interface and implement the interface method in "PdfServiceImpl"

File: web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java [hunk 8]
```diff
@@ -375,6 +409,198 @@ private boolean isEnrollmentCapEnforced(StudyBean currentStudy,StudyBean siteStu
...
+				    	 String studyOID,
+				         String studySubjectIdentifier,
+				         String fullFinalFilePathName) throws IOException {
```
Comment (svadla-oc) [inline web/src/main/java/org/akaza/openclinica/service/StudyParticipantServiceImpl.java]: delete studyOID and studySubjectIdentifier arguments

**Comment by Taoli2018** [comment]
Please see: 
https://github.com/OpenClinica/OpenClinica/pull/2449

**Pull request closed without merging**